# Rapport TP1 Métaheuristiques d'optimization 

### Introduction:

Ce TP consiste à comparer la performance et l'efficacité de trois métaheuristiques du type Monte-Carlo. 
J'ai essayé de regrouper tout le code dans un seul notebook même si le développement de chaque algorithme est fait dans un fichier séparé accompagnant ce rapport. 
Ce notebook représente le rapport du travail et comporte tout le code source et le résultat de l'exécution avec des commentaires et des explications.
Pour les deux algorithmes du hill climbing et du recuit simulé, il se peut que l'algorithme rentre dans une boucle de très longue durée lors de la génération du nouveau candidat à cause des contraintes, il serait utile donc d'interrompre le Kernel et réexécuter l'appel de l'algorithme.

Deux bibliothèques seulement sont utilisées pour ce travail :

In [1046]:
import numpy as np
import pandas as pd
#from tqdm import tqdm 

La fonction 'bornes' permet de retourner les bornes des variables du problème et parce que les bornes des variables x3 et x4 dépendent des valeurs de z1 et z2, on doit calculer les bornes à chaque fois en fonction du nouveau point qu'on a.

Ces bornes n'étaient pas explicites, il fallait donc simplifier les différentes inégalités qu'on a dans le problème. Dans ce qui suit la démonstration :

•	x1 <= 99 : c'est une donnée, mais j'ai choisi comme borne inferieure 1 après plusieurs exécutions ou l'algorithme tournait dans une boucle très longue pour trouver un candidat acceptable.

•	x2 >= 1 : donnée, et j'ai choisi comme borne supérieure 100 pour la même raison

•	0.00954*x3 <= z2 & 0.0193*x3 <= z1 & x3 <= 200 ce qui implique que x3 <= min(0.625*x[1]/0.00954,0.625*x[0]/0.0193, 200), et j'ai choisi comme borne inferieure 1

•	-pi*x3**2* x4 - 4/3*pi*x3**3 <= -1296000  & x4 >=10 implique que x4 >= 1296000/pi*x3**2 - 4/3*x3 (x3!=0) & x4 >= 10

ce qui implique que : x4 >= max(1296000/pi*x3**2 - 4/3*x3, 10) et on a x3 <= 240


In [1047]:
def bornes(x):
    return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])

La fonction objective :

In [1048]:
def cout_soudure(x):
            return 1.7781*0.625*x[1]*x[2]**2 + 0.6224*0.625*x[0]*x[2]*x[3] + 3.1661*(0.625*x[0])**2*x[3] + 19.84*(0.625*x[0])**2*x[2]


Cette fonction permet de vérifier si le nouveau point génère respecte les contraintes ou pas en se servant de la fonction 'bornes'. 

In [1049]:
def contraintes(x):
    _bornes = bornes(x)
    for i in range(len(x)):
        if x[i] < _bornes[i,0] or x[i] > _bornes[i,1]:
            return False
    return True


La fonction d'initialisation du premier point pour les deux algorithmes du hill climbing et du recuit simulé et qui le génère en respectant les contraintes imposées en se servant de la fonction 'bornes'. 

In [1050]:
def initialiser_points():
    x = np.zeros(4)
    _bornes = bornes(x)
    while not contraintes(x):
            x[0]= np.random.uniform(low =  _bornes[0,0],high = _bornes[0,1])
            x[1]= np.random.uniform(low =  _bornes[1,0], high =  _bornes[1,1])
            _bornes = bornes(x)
            x[2]= np.random.uniform(low =  _bornes[2,0],high =  _bornes[2,1])
            _bornes = bornes(x)
            x[3]= np.random.uniform(low =  _bornes[3,0], high =  _bornes[3,1])

    return x



Cette fonction est pour générer une liste de n points pour l'algorithme aléatoire.

In [1051]:
def generer_points(n):
    x = np.zeros([n,4])
    for i in range(n):
            _bornes = bornes(x[i])
            
            while not contraintes(x[i]):
                
                    x[i,0]= np.random.uniform(low =  _bornes[0,0],high = _bornes[0,1])
                    x[i,1]= np.random.uniform(low =  _bornes[1,0], high =  _bornes[1,1])
                    _bornes = bornes(x[i])
                    x[i,2]= np.random.uniform(low =  _bornes[2,0],high =  _bornes[2,1])
                    _bornes = bornes(x[i])
                    x[i,3]= np.random.uniform(low =  _bornes[3,0], high =  _bornes[3,1])

    return x
   


La fonction 'mutation' est utilisée aussi pour les deux algorithmes du hill climbing et du recuit simulé. Elle permet de modifier une des variables choisies aléatoirement du candidat actuel. Elle retourne un nouveau point en fonction de l'ancien à condition qu'il respecte les contraintes.
Pour ce, j'ai utilisé une mutation trouvée dans une référence du cours qui se sert de la distribution gaussienne. En plus, si la fonction ne trouve pas un nouveau candidat acceptable pendant dix itérations, elle change la dimension à modifier.


In [1052]:
def mutation(candidat):
    _bornes = bornes(candidat)

    
    temp_array = candidat

    cpt=0
    while True: #do-while
        
        if(cpt%10 == 0):
            i = np.random.randint(0,4)
            temp=candidat[i]
        
        temp = candidat[i] + np.random.normal(0,1) * 0.01 * (_bornes[i,1] - _bornes[i,0]) 
        temp_array[i] = temp
        
       

        if contraintes(temp_array): 
            candidat[i] = temp
            break
        
        cpt+=1
    return candidat


Les différentes heuristiques de refroidissement pour le recuit simulé:

In [1053]:
def calculer_temperature_expo(t, T_init,epsilon_t=0.025): #exponentiel
    return (1-epsilon_t)**t*T_init



In [1054]:
def calculer_temperature_log(t, T_init): # logarithmique
    if t<3:
        return T_init
    return T_init/np.log(t)


In [1055]:
def calculer_temperature_poly(t,T_init,t_max=200, alpha=2): # polynomial
    return (1-t/t_max)**alpha * T_init # t_max est l'iteration maximale pour laquelle la temperature doit etre nulle


Fonction de calcul de la probalbilite d'acceptation d'un nouveau point dont le score est plus mauvais que celui du point atuel.

In [1056]:
def proba_acceptance(T,deltaE):
    return np.exp(-deltaE/T)

'print_results' permet d'afficher les meilleurs résultats trouvés. 

In [1057]:
def print_results(hist):
    
    scores = hist[:,1]
    best_score = min(scores)
    index = np.where(scores == best_score)
    #print(f"index : {index}")
    print(f"meilleurs dimensions : \n z1 = {0.625*hist[index[0][0],3]} \n z2 = {0.625*hist[index[0][0],4]} \n x3 = {hist[index[0][0],5]} \n x4 = {hist[index[0][0],6]}\n")
    print(f"meilleur score : {best_score}\n")

   



Cette fonction prépare les données pour la présentation dans un tableau a l'aide de la bibliothèque 'pandas'. Elle permet de convertir une liste numpy en une DataFrame, avec le calcul de z1 et z2 à partir de x1 et x2.

In [1058]:
def preparation_table(hist):
    hist[:,3] = [0.625 * a for a in hist[:,3]]
    hist[:,4] = [0.625 * a for a in hist[:,4]]

    table = pd.DataFrame(data=hist[:,1:], columns=["Score","Nbr itrs", "z1","z2","x3","x4"], index=hist[:,0]) 
    
    return table




Fonction pour l'affichage de l'historique détaillé des points et scores.

In [1059]:
def print_table(hist):
      print("historique detaille :\n")
      return preparation_table(hist)

Fonction pour l'affichage d'une table contenant l'analyse de l'historique à l'aide de la méthode describe() de 'pandas'.

In [1060]:
def print_analyse(hist):
    print("Analyse des donnees generees :\n")
    return preparation_table(hist).describe()


L'algorithme de la recherche aléatoire : 

In [1061]:
def aleatoire(n = 50, itrs=100,epsilon = 0.001, samples_size = 10 ,verbose = False):
    
     
       #iterateur 
    i=0
    historique = np.empty([0,7])
    while(i<n):
        i+=1
        k=0
        l=0
        current_points = generer_points(samples_size)
        scores = np.apply_along_axis(cout_soudure, arr = current_points, axis = 1)
        score = np.min(scores)
        current_point = current_points[np.argmin(scores),]
        best_point = current_point
        best_score = score
        best_before_score = best_score
        
        while(k<itrs):
          k+=1   
          current_points = generer_points(samples_size)
          scores = np.apply_along_axis(cout_soudure, arr = current_points, axis = 1)

          score = np.min(scores)
          current_point = current_points[np.argmin(scores),]          
          
          best_before_score = best_score

          if(score < best_score):
                best_score = score
                best_point = current_point
          

          if best_before_score - best_score < epsilon :
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              break
          

          if verbose :
                print(f"itr : {k}, minimum : {current_point}, best score : {score}")
        
        historique = np.append(historique,[np.append([i,best_score,k],[best_point])],axis=0)
    return historique

In [1062]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Veuillez reesayer!")
    
hist = aleatoire(itrs = int(iter), epsilon=float(epsilon))

<ipython-input-1047-6b33492640bc>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [1063]:
print_results(hist)

meilleurs dimensions : 
 z1 = 1.3767210474805958 
 z2 = 1.939598125313187 
 x3 = 60.0784197387637 
 x4 = 108.86495743316941

meilleur score : 20964.93791479527



In [1064]:
print_table(hist)

historique detaille :



Score  Nbr itrs         z1         z2          x3          x4
1.0    26446.032531      10.0   1.963551   1.233420   56.208493  188.101222
2.0    82574.232496      10.0   5.043144   2.109175   89.348620   20.914695
3.0    27213.196844      10.0   2.311289   1.469481   74.252881   39.903838
4.0    71928.021815      10.0   3.010379  11.104022   49.110910  128.222285
5.0    44269.077448      10.0   3.282426   2.989003   42.422749  212.249387
6.0   264220.702670      10.0   3.957039  11.552569   92.223019  219.952712
7.0    82847.259180      10.0   2.059841   4.511283   92.772015   45.322374
8.0    27497.102975      10.0   2.009807   3.970352   38.092822  234.950957
9.0   103060.435245      10.0   1.246734  14.609055   60.445743  121.309502
10.0   24153.032737      10.0   0.964143   6.403787   38.980109  232.882345
11.0   73633.561229      10.0   3.145237  12.329053   43.927577  193.584982
12.0   96339.731528      10.0   4.290667  15.996087   41.608579  188.308716
13.0  185059.264446      10.0   9.290813   0.790988   79.820741   53.614337
14.0  166542.045576      10.0   2.684896  44.684154   41.911959  225.926258
15.0   20964.937915      10.0   1.376721   1.939598   60.078420  108.864957
16.0  132762.763654      10.0   3.722426  28.187951   42.139461  227.394171
17.0  134683.736980      10.0   5.083763   3.501446   70.423354  222.206549
18.0   40648.304925      10.0   2.919902   2.823971   48.508654  179.133740
19.0   22321.414127      10.0   1.786925   0.671907   52.309133  230.476319
20.0  131822.765347      10.0   2.711935   4.296150  118.083729   36.281928
21.0   87686.604133      10.0   1.919159  13.636884   53.101006  205.548072
22.0   46002.526873      10.0   1.197896  11.395279   42.270056  238.388301
23.0  141416.338862      10.0   4.790566  19.896983   49.432785  147.497018
24.0  134205.939041      10.0   5.674498   9.645079   53.802345  171.905186
25.0  215597.472232      10.0   3.395878  13.808384   85.618355   73.686546
26.0   46493.901462      10.0   2.957145   3.193032   57.756821  130.860787
27.0   93651.651172      10.0   5.978680   3.908220   39.991455  206.793959
28.0  111457.804940      10.0   2.995099  16.050647   56.217028   84.513058
29.0  109649.751162      10.0   2.101373   4.741769   91.608031  230.712337
30.0   69886.565661      10.0   2.339200   4.200811   81.476330   84.269023
31.0  124786.306025      10.0   5.103938   5.425159   67.658123  153.528140
32.0   51753.914360      10.0   1.107816  15.227563   40.064386  232.210806
33.0   60495.974203      10.0   2.011319  11.473695   44.860392  229.656731
34.0  102555.266011      10.0   1.903332  29.698717   40.744872  200.193211
35.0   57877.177000      10.0   1.911155  11.710926   45.915515  160.917905
36.0   75845.520526      10.0   3.971309   8.946315   43.712719  201.112234
37.0   34359.676952      10.0   2.988588   1.668332   38.136266  234.715243
38.0  220279.851647      10.0   1.787678  42.655289   52.269392  142.794531
39.0  157646.414929      10.0   8.338462   6.936843   57.219377   74.125464
40.0  136521.494891      10.0   3.298020   7.455259   78.480167  194.028258
41.0   43533.229999      10.0   2.066090   2.436037   82.608741   58.272453
42.0  165777.277817      10.0   5.725873  23.272309   49.901546  107.490618
43.0   28894.108989      10.0   2.069151   1.674991   53.855742  189.127181
44.0   52751.470436      10.0   2.108554   4.796672   67.390155   78.751612
45.0  271784.532353      10.0  12.294765   2.372412   43.694487  163.218825
46.0   25859.933344      10.0   1.580747   2.536325   54.742797  155.917138
47.0  116204.437000      10.0   1.107723  31.233315   43.980034  225.501549
48.0  195083.043319      10.0   8.873900  13.687644   51.133001   96.971997
49.0   79675.288631      10.0   2.557032   5.594523   79.032649   49.757644
50.0   33511.595687      10.0   1.872268   1.597699   68.596442  168.874620

In [1065]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2          x3          x4
count      50.000000      50.0  50.000000  50.000000   50.000000   50.000000
mean    97004.653787      10.0   2.111111   6.325632   58.958800  156.018834
std     64174.407375       0.0   1.451910   6.441052   18.637848   67.048378
min     20964.937915      10.0   0.602589   0.419942   38.092822   20.914695
25%     44702.439804      10.0   1.206411   1.790768   43.766434   99.601652
50%     82710.745838      10.0   1.686510   3.749472   53.451675  170.389903
75%    133845.145194      10.0   2.479839   8.546846   69.966626  218.026881
max    271784.532353      10.0   7.684228  27.927596  118.083729  238.388301

L'algorithme du Hill Climbing : 

In [1066]:
def hill_climbing(itrs = 100,n = 50, epsilon = 0.001, verbose = False):
    i=0
    historique = np.empty([0,7])

    while(i<n):
      i+=1
      currentPoint = initialiser_points()
      bestScore = cout_soudure(currentPoint)
      bestPoint = currentPoint
      beforeBestScore = bestScore
      k=0 #iterateur 
      l=0

      while(k<itrs):

          k+=1   
          currentPoint = mutation(currentPoint) 
          score = cout_soudure(currentPoint)
          
          beforeBestScore = bestScore

          if(score < bestScore):
                bestScore = score
                bestPoint = currentPoint

          if beforeBestScore - bestScore < epsilon :
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              break
          

          if verbose :
                print(f"itr : {k}, minimum : {currentPoint},  score : {score}")
    
      historique = np.append(historique,[np.append([i,bestScore,k],[bestPoint])],axis=0)

    return historique

In [1067]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Veuillez reesayer!")
hist = hill_climbing(itrs = int(iter), epsilon=float(epsilon))

<ipython-input-1047-6b33492640bc>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [1068]:
print_results(hist)

meilleurs dimensions : 
 z1 = 6.244385214150605 
 z2 = 22.640096624264018 
 x3 = 51.41133320591536 
 x4 = 237.40994015593253

meilleur score : 155288.89816842816



In [1069]:
print_table(hist)

historique detaille :



Score  Nbr itrs         z1         z2          x3          x4
1.0   1.360817e+07      10.0  61.512783   2.757661  166.057511   73.344507
2.0   1.318264e+06      10.0  12.877405  50.325956   97.069701  127.850539
3.0   5.726143e+06      10.0  52.655890  46.837045   75.276884  102.174104
4.0   4.820925e+06      10.0  56.175399  32.031550   66.196113   42.010798
5.0   4.661274e+06      10.0  33.288750  47.794660  112.638772  229.987234
6.0   5.673072e+06      10.0  31.278239  12.739242  187.959317  188.643330
7.0   7.041448e+06      10.0  46.945765  21.496941  130.269742   64.602038
8.0   4.752305e+06      10.0  44.504272   7.636775   92.039076  137.686750
9.0   7.029334e+06      10.0  47.443259  19.827643  119.821431  109.922664
10.0  5.101117e+05      10.0   4.657935  43.617987   75.700142  114.813393
11.0  1.482733e+07      10.0  53.486439  32.600512  199.742490   85.856533
12.0  6.573209e+05      10.0  12.878949  32.291186   62.475679  221.858005
13.0  2.731895e+06      10.0  34.562508   3.064138   87.852019  107.146894
14.0  5.387905e+05      10.0  13.770843  33.097098   66.484769   45.975030
15.0  4.260787e+06      10.0  48.761138  31.199806   51.263092  186.795935
16.0  3.512838e+06      10.0  17.571364  59.642383  146.579691  130.353573
17.0  2.377627e+06      10.0  39.341324  22.144425   44.286324  158.097478
18.0  7.779045e+05      10.0   9.339378  33.161207   97.942760  130.988971
19.0  4.087260e+06      10.0   6.160579  62.424836  191.798557  103.326026
20.0  1.388630e+07      10.0  54.290275   3.225031  199.770742  135.895880
21.0  7.324839e+06      10.0  51.265399  30.721874   98.146228  161.724901
22.0  1.479164e+06      10.0  21.373052  19.178555   85.768349  174.310711
23.0  8.984011e+06      10.0  51.367697  15.719149  135.981767  107.940730
24.0  7.152813e+06      10.0  32.306355  51.919367  166.948749  226.290396
25.0  1.470883e+06      10.0  13.365877  51.822570  104.766494  103.077508
26.0  7.245072e+05      10.0   3.105116  55.633024   87.319843  112.048501
27.0  1.552889e+05      10.0   6.244385  22.640097   51.411333  237.409940
28.0  4.385073e+06      10.0  37.444577  56.981115   95.262016  126.732080
29.0  3.791586e+05      10.0   4.018719   8.626361  133.110574  168.542150
30.0  1.425222e+06      10.0  17.180869  37.174121   85.492490  238.765568
31.0  1.272124e+07      10.0  56.648304  50.302633  150.350674   73.647629
32.0  3.902818e+05      10.0   2.475636  25.235355   91.768216  147.973427
33.0  1.255466e+06      10.0  21.816151  36.919922   71.780812   96.476629
34.0  1.050898e+06      10.0  20.583725  48.636875   48.188999  231.721689
35.0  2.707874e+06      10.0  19.933002  26.308957  163.927841   48.168588
36.0  2.463623e+06      10.0  19.231615  18.591066  158.928094  174.198680
37.0  5.282156e+06      10.0  41.701323  22.775212  128.507100   51.720035
38.0  6.520714e+06      10.0  59.801290  38.126679   65.671242  122.423559
39.0  5.204399e+06      10.0  49.175995  19.535004   60.128041  239.925620
40.0  3.153211e+06      10.0  26.830644  36.910506  133.682266   15.740485
41.0  6.988855e+06      10.0  40.192698  24.646226  136.930348  210.538426
42.0  6.093032e+05      10.0   8.821789  37.427068   89.011843   39.531057
43.0  1.249078e+06      10.0   9.988707  41.779609  113.384252   68.163627
44.0  2.780114e+06      10.0  28.211453  47.979434   88.635670  208.568735
45.0  1.248224e+06      10.0  15.564474   8.729709  132.201209  193.580867
46.0  3.324313e+06      10.0  23.368796  56.919647  129.945203   57.312741
47.0  8.003893e+06      10.0  45.883074   6.429762  177.893138   30.750978
48.0  5.455889e+06      10.0  46.962375  30.919856   97.390763   73.111587
49.0  7.182301e+06      10.0  51.981270  17.211673   85.176496  215.222072
50.0  5.889294e+05      10.0   5.469611  45.348755   81.275313  138.780547

In [1070]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2          x3          x4
count  5.000000e+01      50.0  50.000000  50.000000   50.000000   50.000000
mean   4.289216e+06      10.0  18.922706  19.863328  110.404203  131.834583
std    3.745407e+06       0.0  11.562295  10.328251   42.335135   63.999823
min    1.552889e+05      10.0   1.547273   1.723538   44.286324   15.740485
25%    1.250675e+06      10.0   8.416949  12.255103   82.250609   76.699855
50%    3.418576e+06      10.0  18.590529  20.100855   97.666761  127.291310
75%    6.322071e+06      10.0  29.576899  29.040608  135.406892  183.674629
max    1.482733e+07      10.0  38.445490  39.015523  199.770742  239.925620

L'algorithme du recuit simulé : 
Après chaque m itérations, on ajoute un nombre aléatoire par la loi uniforme entre 0 et 1. Par défaut, m est fixe à 10 itérations.


In [1071]:
def recuit_simule(T_init = 400,temperature= calculer_temperature_poly,itrs= 100,n = 50, m=10,epsilon = 0.001, verbose = False):
      # T_init ne doit pas etre posee comme ça
      
    
    historique = np.empty([0,7])
    i=0
    while(i<n):
        i+=1
        current_point = initialiser_points()
        best_point = current_point
        current_score = cout_soudure(current_point)
        best_score = current_score
        bestc_before_score = best_score
        k=0 # iterateur 
        l=0 # iterations pour suivre l'amelioration
        while(k<itrs):
          k+= 1
          
          new_point = mutation(current_point)
          new_score = cout_soudure(new_point)
          
          deltaE = new_score - current_score
          
          best_before_score = best_score

          if  deltaE <= 0 :   
              current_point = new_point
              current_score = new_score
              if current_score < best_score:
                  best_point = current_point
                  best_score = current_score
                
          else :
              T = temperature(k,T_init) + k//m * np.random.uniform(0,1) # Rechauffement apres chaque m iterations
              if np.random.uniform(0,1) < proba_acceptance(T,deltaE):
                  current_point = new_point
                  current_score = new_score
  

          if best_before_score - best_score < epsilon :  #peu ou pas d'amelioration 
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              current_point = best_point         #Retour au dernier meilleur point
              current_score = best_score
          

          if verbose :
                print(f"itr : {k}, point : {current_point}, score : {current_score}")
                
        historique = np.append(historique,[np.append([i,best_score,k],[best_point])],axis=0)

    return historique

Resultats pour le refroidissement polynomial : 

In [1072]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(itrs = int(iter), epsilon=float(epsilon))

<ipython-input-1047-6b33492640bc>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [1073]:
print_results(hist)

meilleurs dimensions : 
 z1 = 6.051209621058784 
 z2 = 10.803206806083788 
 x3 = 65.57144963126603 
 x4 = 218.92439493477983

meilleur score : 205109.66857755402



In [1074]:
print_table(hist)

historique detaille :



Score  Nbr itrs         z1         z2          x3          x4
1.0   1.045029e+07      10.0  40.444984  40.659709  193.936377  145.707198
2.0   6.399583e+06      10.0  41.185731  50.763870  119.058380  141.068556
3.0   2.171174e+06      10.0  24.774792  42.178508   93.266460  119.811462
4.0   4.897634e+06      10.0  38.143778  27.107329  135.821421   25.499086
5.0   7.550020e+05      10.0   2.926296  30.044106  119.757988   50.793648
6.0   9.296921e+06      10.0  57.058660  33.944276   94.320470  195.317029
7.0   6.060461e+06      10.0  36.955314  15.989191  174.578577   57.313538
8.0   5.420722e+06      10.0  48.420753  50.394747   72.150098  177.309242
9.0   1.024110e+07      10.0  45.844159  48.677662  172.589338   82.013695
10.0  8.532889e+06      10.0  48.966956  22.821805  125.321643  175.606646
11.0  1.014315e+07      10.0  58.281058  54.640566  121.256242   44.626499
12.0  1.730081e+06      10.0  30.793587  11.852765   54.458711  158.931519
13.0  1.842197e+06      10.0  14.590712  18.539851  181.626054   17.952481
14.0  1.198730e+06      10.0  18.012148  16.960618  114.242301   30.232082
15.0  2.021178e+06      10.0  18.976605  30.384158  138.787771   31.463642
16.0  8.821682e+06      10.0  49.461963  38.242134  116.930347  195.351112
17.0  1.202548e+06      10.0  25.550028  43.347155   54.140636  123.991534
18.0  4.166630e+06      10.0  25.421963  34.734276  170.172509   87.116065
19.0  3.549836e+06      10.0  23.125587  56.460611  136.562802  121.213473
20.0  1.434387e+06      10.0   6.089481  43.716065  128.423455  171.231067
21.0  1.390617e+07      10.0  49.981192  33.384660  192.979726  173.892781
22.0  7.277801e+06      10.0  56.236464  58.261894   91.859824   48.375087
23.0  1.635748e+07      10.0  58.539027  43.635130  165.380624  226.503168
24.0  9.854490e+05      10.0  20.169052  26.768766   71.915813  189.958405
25.0  1.361519e+07      10.0  56.907859  27.013614  142.412276  231.355515
26.0  9.045133e+06      10.0  61.305425  39.244781   89.967720  155.476153
27.0  2.828172e+06      10.0  40.355655  58.605178   57.058819  118.290527
28.0  2.983553e+06      10.0  32.831424  57.636606   77.478254  159.542262
29.0  2.179531e+06      10.0   5.419387  58.854758  141.722916   43.247648
30.0  1.095667e+07      10.0  47.483016  45.812884  156.745490  176.547802
31.0  4.909491e+06      10.0  27.177798  49.376553  168.879134   31.636706
32.0  1.105269e+07      10.0  36.541030  61.952066  194.527406  238.265927
33.0  6.246029e+06      10.0  46.644952  47.587670  119.822493   31.556758
34.0  3.335034e+06      10.0  41.928504  28.012988   67.282849  135.339404
35.0  7.430755e+06      10.0  33.024149  38.370704  177.321055  221.817268
36.0  1.154033e+06      10.0  20.914173  57.384896   61.958060  108.993942
37.0  1.752066e+06      10.0  14.375938  11.892522  199.947280   44.484261
38.0  4.928374e+06      10.0  55.322430  28.499254   41.216824  210.592347
39.0  2.497507e+05      10.0   2.364010  61.331602   47.189285  222.838230
40.0  3.139537e+05      10.0   7.985973  62.442847   42.182605  238.367261
41.0  1.503622e+06      10.0   6.747941  40.356250  140.438885   91.635278
42.0  5.258842e+05      10.0   5.542623  12.806513  139.767047  151.742311
43.0  1.097768e+06      10.0   5.909544  58.049942   97.656219  157.178024
44.0  8.620730e+06      10.0  49.864966  32.772231  146.181177   13.218691
45.0  6.436590e+06      10.0  44.251078  17.288647  111.422276  212.643724
46.0  3.287858e+06      10.0  18.878313  50.680245  148.329676  156.554859
47.0  4.324525e+06      10.0  52.872388   1.131473   49.326062  151.220243
48.0  5.156201e+06      10.0  26.393014  27.781079  181.528441  196.558087
49.0  2.051097e+05      10.0   6.051210  10.803207   65.571450  218.924395
50.0  9.070101e+05      10.0  22.447346  39.481160   42.874558  212.367904

In [1075]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2          x3          x4
count  5.000000e+01      50.0  50.000000  50.000000   50.000000   50.000000
mean   5.078177e+06      10.0  20.118630  23.733494  118.966916  134.433491
std    4.108354e+06       0.0  11.267012  10.057414   47.845463   70.251724
min    2.051097e+05      10.0   1.477506   0.707171   41.216824   13.218691
25%    1.560236e+06      10.0  11.814304  17.047354   73.482137   63.488577
50%    4.245577e+06      10.0  20.579866  24.601856  120.539368  151.481277
75%    8.257356e+06      10.0  30.116449  31.630544  154.641536  193.977373
max    1.635748e+07      10.0  38.315891  39.026779  199.947280  238.367261

Resultats pour le refroidissement exponentiel : 

In [1076]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(temperature = calculer_temperature_expo,itrs = int(iter), epsilon=float(epsilon))

<ipython-input-1047-6b33492640bc>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [1077]:
print_results(hist)

meilleurs dimensions : 
 z1 = 1.5666855005764102 
 z2 = 1.4619529692213185 
 x3 = 78.79538203279509 
 x4 = 236.98973914469838

meilleur score : 33569.129424975254



In [1078]:
print_table(hist)

historique detaille :



Score  Nbr itrs         z1         z2          x3          x4
1.0   2.300645e+06      10.0  37.046451  21.966084   44.432158  196.210233
2.0   2.993442e+06      10.0  12.682936  45.753551  168.052232   95.786368
3.0   7.989012e+06      10.0  34.139303  53.201009  167.474103  208.525453
4.0   6.079104e+06      10.0  53.630260  22.151267   74.401049  190.381696
5.0   1.219749e+06      10.0   9.361609  17.664856  160.234735  119.381867
6.0   5.543292e+06      10.0  37.153816  17.955182  167.426908   54.603844
7.0   3.913124e+06      10.0  36.496789  61.397690   92.199682   98.748188
8.0   1.389267e+06      10.0  13.795938  16.573594  143.052134  191.028324
9.0   7.469613e+06      10.0  49.731573  12.635006  146.377878   47.950934
10.0  9.704645e+05      10.0  21.041561  18.716166   57.440169  202.403447
11.0  7.123384e+06      10.0  33.438395  48.333677  161.946453  230.711149
12.0  8.609436e+06      10.0  41.032398  51.788738  168.218932   49.859273
13.0  3.593955e+06      10.0  34.554973  29.672943  105.160671   91.381086
14.0  1.559533e+07      10.0  57.120197  37.500627  177.656682  135.283987
15.0  3.654740e+06      10.0  34.728439  14.549744  132.484361   29.266382
16.0  3.417131e+06      10.0  22.077071  36.506232  153.198676  112.997395
17.0  4.889099e+05      10.0   6.785665  34.454699   76.087689  209.852685
18.0  1.298848e+07      10.0  61.512521  12.921873  162.575879   44.837138
19.0  1.689567e+07      10.0  57.186675  11.524981  193.867473  207.001761
20.0  9.382186e+05      10.0  20.184642  25.289851   59.564135  145.816815
21.0  2.654846e+06      10.0  38.794024  28.542962   65.086709   88.108702
22.0  2.218928e+06      10.0   6.579609  33.570580  177.003614  238.147998
23.0  1.493709e+06      10.0  22.285792  18.082506   96.781974   81.953953
24.0  3.492776e+05      10.0   3.026717   7.508611  153.448428   22.063811
25.0  2.538777e+06      10.0  36.347161  22.124240   68.668662  114.030875
26.0  7.729626e+06      10.0  48.883321  22.850582  117.647411  147.749519
27.0  8.057558e+05      10.0  15.583130  29.060149   79.100020   65.991573
28.0  5.610951e+06      10.0  45.929794  51.741449   82.415207  171.889978
29.0  1.512141e+06      10.0  17.987312  23.432146  118.092397  119.087657
30.0  7.003882e+06      10.0  57.443454  17.957731   66.291408  199.878524
31.0  3.913685e+06      10.0  27.172710  44.079366  127.191769  174.318436
32.0  5.679241e+06      10.0  18.754716  62.335048  197.135768  154.710849
33.0  1.426767e+07      10.0  61.419052  59.976195  136.785049  162.384255
34.0  1.651715e+06      10.0  14.027181  17.188187  162.731601  129.321134
35.0  3.986951e+06      10.0  36.277267  11.770374  119.354631  136.198353
36.0  3.356913e+04      10.0   1.566686   1.461953   78.795382  236.989739
37.0  1.110916e+06      10.0  18.162528  49.118950   79.644525   19.389830
38.0  1.594129e+07      10.0  54.206192  29.641541  186.720285  212.101084
39.0  6.372201e+06      10.0  26.471368  61.402385  180.855370   69.000227
40.0  9.590987e+06      10.0  42.248248  44.841111  166.158487  155.532001
41.0  1.986331e+06      10.0  19.871624  33.063194  130.074890   55.822648
42.0  1.370036e+07      10.0  59.010848  38.915520  167.049208   26.358060
43.0  3.161058e+06      10.0  29.828632  17.382335  120.956154  120.436470
44.0  3.759676e+06      10.0  15.307013  42.984370  191.376600   27.591157
45.0  1.068481e+06      10.0  14.186334  37.062459  103.736898   46.335828
46.0  4.342666e+06      10.0  49.078811   6.340272   65.209706  132.188578
47.0  2.733006e+06      10.0   9.312883  42.009341  178.570597   33.478557
48.0  1.468514e+07      10.0  58.335115   7.105078  165.365359  239.119552
49.0  1.633991e+05      10.0  15.053293   1.717006  176.600295   23.352710
50.0  8.677899e+06      10.0  43.441547  21.279047  158.819396  192.482015

In [1079]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2          x3          x4
count  5.000000e+01      50.0  50.000000  50.000000   50.000000   50.000000
mean   5.238342e+06      10.0  19.753670  18.438781  130.590396  125.160842
std    4.665248e+06       0.0  10.964710  10.431272   44.583899   68.900597
min    3.356913e+04      10.0   0.979178   0.913721   44.432158   19.389830
25%    1.547035e+06      10.0  10.115110  10.908103   84.861325   58.364879
50%    3.707208e+06      10.0  21.118031  16.822754  139.918591  124.878802
75%    7.383056e+06      10.0  28.317333  26.712883  167.332483  190.866667
max    1.689567e+07      10.0  38.445325  38.959405  197.135768  239.119552

Resultats pour le refroidissement logarithmique : 

In [1080]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(temperature = calculer_temperature_log,itrs = int(iter), epsilon=float(epsilon))

<ipython-input-1047-6b33492640bc>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[1,99],[1,100],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [1081]:
print_results(hist)

meilleurs dimensions : 
 z1 = 8.23389282173043 
 z2 = 4.6533976492847575 
 x3 = 56.81696257998691 
 x4 = 216.53680614608822

meilleur score : 212236.5168149154



In [1082]:
print_table(hist)

historique detaille :



Score  Nbr itrs         z1         z2          x3          x4
1.0   3.044398e+06      10.0  21.733553  42.311529  128.845632  181.559850
2.0   5.819002e+06      10.0  27.204209  46.184176  164.101668  236.167178
3.0   3.593838e+06      10.0  24.968947  41.217912  145.136135   97.865534
4.0   8.052063e+06      10.0  55.454446   9.171618  101.326605  129.357000
5.0   6.519317e+05      10.0  17.299597  29.760628   51.161476  173.575762
6.0   1.301384e+07      10.0  58.315571  40.512082  148.831318   85.137100
7.0   3.537250e+06      10.0  36.327623  56.966349   97.793076   33.474073
8.0   6.453829e+05      10.0  14.104424  28.556903   72.736654  178.936080
9.0   2.122365e+05      10.0   8.233893   4.653398   56.816963  216.536806
10.0  7.924082e+05      10.0   6.946065  44.805888   91.648194  141.188756
11.0  4.204817e+05      10.0  11.588788   8.405639   85.232631   88.002058
12.0  2.970747e+05      10.0   8.056117  40.475487   53.571417  133.796897
13.0  1.385699e+06      10.0  28.922784  38.633798   49.775720  141.729498
14.0  2.535233e+06      10.0  13.159530  51.790045  143.667649  171.806337
15.0  8.896318e+05      10.0  15.576535  45.252993   76.559256   32.750950
16.0  6.230358e+06      10.0  24.190646  44.788281  195.054868  220.118871
17.0  8.882462e+06      10.0  39.463333  15.546001  186.678902  230.527918
18.0  2.759852e+05      10.0   6.946089  48.439087   44.602775  179.775767
19.0  7.478919e+06      10.0  49.021651  35.155970  134.629641   17.044307
20.0  5.005246e+06      10.0  42.331171  38.685975   78.112613  233.912829
21.0  7.563811e+06      10.0  45.236630  35.042067  145.411426   45.624234
22.0  1.003099e+06      10.0  11.902761  48.342285   94.535913   13.626230
23.0  6.329107e+06      10.0  38.943167   4.407041  168.456683  116.839420
24.0  3.438843e+06      10.0  42.736044   1.289748   54.113261  204.272732
25.0  7.204454e+06      10.0  54.615373  13.640309  101.329118   99.578073
26.0  4.296027e+06      10.0  38.605958  60.827273   76.114555  228.556425
27.0  1.255535e+07      10.0  52.446607  50.991792  173.912573   36.697296
28.0  9.342529e+06      10.0  47.873928   3.756564  145.192696  224.436576
29.0  5.523499e+06      10.0  55.492562  36.554865   46.441730  229.920573
30.0  1.390417e+07      10.0  60.086421  51.751446  163.073843   25.298982
31.0  1.157429e+07      10.0  51.181044  47.313948  165.140657   76.948014
32.0  1.833070e+07      10.0  60.987156  57.314714  161.067329  231.316870
33.0  6.992335e+06      10.0  60.967149  25.066256   75.606762   88.862761
34.0  1.041349e+06      10.0   4.715480  14.813101  190.942603   81.216361
35.0  4.608189e+06      10.0  28.951276  31.639056  142.947509  220.481276
36.0  3.315158e+06      10.0  45.430667   1.345841   40.040752  218.043215
37.0  9.378929e+05      10.0  15.791157  53.675266   73.394379   51.134536
38.0  1.180376e+06      10.0   3.083237  48.865827  113.313339  176.116026
39.0  3.136240e+06      10.0  43.487299  54.594606   61.792313   60.783226
40.0  1.498135e+06      10.0  24.563153  49.547654   63.166392  135.775757
41.0  1.951424e+06      10.0  19.783500  41.182523   94.802383  231.518705
42.0  8.337181e+05      10.0   4.086105  19.862387  141.612504  210.838735
43.0  4.805076e+06      10.0  26.654153  44.060713  173.254366   17.613465
44.0  4.643287e+06      10.0  37.584662  19.350631  112.143828  175.241883
45.0  1.153261e+06      10.0  17.034887  60.181842   69.990710  193.641616
46.0  7.446234e+06      10.0  45.677626   8.728570  137.286401  139.951037
47.0  2.490756e+06      10.0  30.386253  52.101275   86.891684   58.312084
48.0  9.307094e+06      10.0  55.017224  42.767571  118.453275   90.003864
49.0  1.600501e+06      10.0  11.573091  31.932270  132.327150  230.625302
50.0  3.621663e+06      10.0  29.715762  33.733596  131.046733   87.491005

In [1083]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2          x3          x4
count  5.000000e+01      50.0  50.000000  50.000000   50.000000   50.000000
mean   4.687840e+06      10.0  19.680691  21.949935  111.201721  138.480597
std    4.149847e+06       0.0  11.240813  10.897171   44.876050   73.975693
min    2.122365e+05      10.0   1.927023   0.806093   40.040752   13.626230
25%    1.160040e+06      10.0   9.768869  13.227096   73.947475   82.196546
50%    3.565544e+06      10.0  18.333449  25.308615  106.736473  140.569897
75%    7.151424e+06      10.0  28.509929  30.259304  145.178556  215.112288
max    1.833070e+07      10.0  38.116972  38.017046  195.054868  236.167178

### Conclusion

En moyenne sur 50 exécutions des 3 algorithmes, l'algorithme aléatoire est celui qui a donné les meilleurs résultats ce qui est assez surprenant, en choisissant pour toutes les exécutions effectuées 10 itérations par cycle (50 cycles) et 0.01 pour epsilon. 

En conclusion, le meilleur coût trouvé est de 20964.93791479527.
